### DISCUSSION TITLE
Will remix support config-based routing, like useRoutes from react-router?

### DISCUSSION THREAD

**Discussion by joeprivettsmart** [remix-run/remix#4288] [General]
I'm struggling with the logic behind using the file system and naming these files in certain ways to setup your routes. This seems like an anti-pattern as you want your folder structure to be organised into domains that teams can take ownership of and that doesn't necessarily follow the way it will look like in the URL and hence how you set-up your folder/file structure. Appreciate that Remix is providing what feels like escape hatches to get round this, but it feels completely wrong to be using the file-system to define routes.
The `useRoutes` config-based routing seems way more intuitive. for example here:
https://ui.dev/react-router-route-config

Will Remix support config-based routing?

**Comment by sergiodxa** [comment]
Remix already supports that using the routes function in the remix.config.js https://remix.run/docs/en/v1/api/conventions#routes

**Comment by joeprivettsmart** [comment]
In some cases you might want use a ternary to decided what to render, or introduce a redirect based on some business logic. I can't see how this can be done with the `routes` function. If the second argument accepted a callback, then yes this could be achieved nicely, but at the moment it just allows you to enter a file name and I'm assuming that this file needs to export the meta/loader ?

**Reply by kiliman** [reply]
Remix requires routes to be statically defined at compile time.

**Reply by sergiodxa** [reply]
> use a ternary to decided what to render

This belongs on the route component.

> introduce a redirect based on some business logic

You can do that inside a loader.

---

Aside of that, as kiliman said routes needs to be defined statically at compile time, the routes function let you avoid the file system convention and define routes from any other file.
